From 70e88cf7772c455cf623d0027caa266682a9f0a3 Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Mon, 03 May 2021 23:30:25 +0300
Subject: [PATCH] [SW-26273] fit: error response infrastructure

Instead of returning an anemic error value as a result
of a FW protocol response error. A future compatible,
more flexible, and more informative, approach is needed.

As a result of a FWLP error response, a FIT is being
been made by the FW, ready to be fetch by the FWL.
The current FIT used as a response of an error, is identical
to the one used by the proto_status protocol command.

In order to support the legacy code with minor changes.
The code "rides" over the module_fwupd_send_status command.

And the legacy send status command was renamed with the "_fast"
postfix.

Change-Id: I9bf335ba435454659b750c7a6457379dc64343a3
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/preboot/module_fwupd.c b/preboot/module_fwupd.c
index 8c83b25..7279f3f 100644
--- a/preboot/module_fwupd.c
+++ b/preboot/module_fwupd.c
@@ -41,26 +41,149 @@
 
 static struct fwupd_state state;
 
-static void module_fwupd_send_status(enum hl_upd_sts sts, void *offset)
+static void fwupd_get_proto_status(struct hl_upd_proto_status *status, int err, int serr)
+{
+	status->version = HL_UPD_PROTO_VERSION;
+	status->compat_mask = cpu_boot_dev_sts_get_all();
+	status->fw_comp = HL_FW_COMP_PREBOOT;
+	status->err = err;
+	status->sub_err = serr;
+}
+
+static int fwupd_create_resp(void *res_fit, const int load_type,
+		const char *prop_name, const void *prop_data,
+		int *prop_data_sz)
+{
+	int err;
+	unsigned int sz;
+
+	sz = ((*prop_data_sz + FIT_RSP_MARGIN_SIZE_IN_KB) > FIT_DFL_RSP_SIZE_IN_KB) ?
+		(*prop_data_sz + FIT_RSP_MARGIN_SIZE_IN_KB) : FIT_DFL_RSP_SIZE_IN_KB;
+
+	err = fit_init_resp(res_fit, sz);
+	if (err)
+		goto fit_init_err;
+
+	err = fdt_property_u32(res_fit, HL_FIT_LOAD_TYPE_PROP, load_type);
+	if (err)
+		goto fit_init_err;
+
+	err = fdt_property(res_fit, prop_name, prop_data, *prop_data_sz);
+	if (err)
+		goto fit_init_err;
+
+	err = fit_fin_resp(res_fit, &sz);
+	if (err)
+		goto fit_init_err;
+
+	*prop_data_sz = sz;
+
+fit_init_err:
+	return err;
+}
+
+static void module_fwupd_send_status_fast(enum hl_upd_sts sts, void *offset)
 {
 	struct hl_upd_status status;
 	uintptr_t off = (uintptr_t)offset;
-	u32 err, serr;
-
-	/*
-	 * reads the errors and clear the error registration
-	 * then printing it out
-	 * it is a preparation code
-	 */
-	if (preboot_error_read_and_clr(&err, &serr)) {
-		preboot_puts_hex_raw("STS: ERR = ", err, 0);
-		preboot_puts_hex(", SUB-ERR = ", serr, 0);
-	}
 
 	status.val = 0;
 	status.status = sts;
 	status.offset = (u32)(off & ~HL_UPD_STS_OFFSET_MASK);
 	regmap_writel(status.val, mmUPD_STS);
+}
+
+static int module_fwupd_proto_status(struct hl_upd_proto_status *ret_status)
+{
+	struct hl_upd_proto_status status;
+	u32 sts_err, sts_serr;
+	int size = sizeof(status);
+	int err;
+
+	state.map_offset = (void *)FIT_IMG_BASE;
+
+	preboot_error_read_and_clr(&sts_err, &sts_serr);
+	fwupd_get_proto_status(&status, sts_err, sts_serr);
+
+	err = fwupd_create_resp(state.map_offset, FIT_LOAD_TYPE_PROTO_STATUS,
+				HL_FIT_PROTO_STATUS_PROP, &status, &size);
+	state.map_size = size;
+
+	if (err)
+		preboot_puts("Failed to create status FIT image\n");
+
+	if (ret_status)
+		*ret_status = status;
+
+	return err;
+}
+
+static void module_fwupd_send_status(enum hl_upd_sts sts, void *offset)
+{
+	int err;
+	void *tmp_offset = offset;
+
+	/*
+	 * can send fast if not error
+	 * if it is error - must be slow
+	 */
+	switch (sts) {
+	case HL_UPD_STS_MAP_ERR:
+	case HL_UPD_STS_ERR:
+	case HL_UPD_STS_RST_ERR:
+	case HL_UPD_STS_VALID_ERR:
+	case HL_UPD_STS_TIMEOUT_ERR:
+		/*
+		 * The error inform function, work in such a way
+		 * that the error will be set only once (assuming no concurrency)
+		 * so if the error inform function is called before entring the
+		 * current function, module_fwupd_send_status, the second call will not
+		 * overrun its values.
+		 * it allows support legacy code by setting the sent error status as protocol
+		 * error and set it's minor value.
+		 * if
+		 */
+		preboot_error_inform(HL_FW_ERR_PROTO, sts);
+		if (state.exec_res != FOUND_SLOW_RETURN) {
+			/*
+			 * error status response requires FIT creation, that implies - SLOW MODE
+			 * if an error status request has been done from within a slow
+			 * mode invocation. The request can be serve and skip this conditional
+			 * block.
+			 * otherwise, the control should be returned to the main event loop,
+			 * with request for change into SLOW MODE and execute a command state to
+			 * send the error status. There is not enough commands bits to add this
+			 * command state, so the alternative for that, is using a flag to ask for
+			 * running the state. That condition is not enough; an error should be
+			 * inform in order to trigger the response.
+			 * Both of those condition should be checked in the module's check and run
+			 * functions.
+			 * The error was set previously (2 lines above), and the request
+			 * to change the state is done by this flag.
+			 *
+			 * WARNING:	don't change state.exec_res it is the check function
+			 *		responsibility
+			 *
+			 */
+			state.sending_error_flag = true;
+			break;
+		}
+
+		/* if an explicit FIT was attached - use it */
+		if (!offset) {
+			tmp_offset = state.map_offset;
+			err = module_fwupd_proto_status(NULL);
+
+			/* if prepare fails - just don't include it*/
+			if (err)
+				tmp_offset = 0;
+		}
+
+	default:
+		/* on STS_OK / STS_ACK etc,  the original offset is set */
+		module_fwupd_send_status_fast(sts, tmp_offset);
+	}
+
 }
 
 void module_fwupd_setup(void)
@@ -350,65 +473,17 @@
 	return err;
 }
 
-static int fwupd_create_resp(void *res_fit, const int load_type,
-		const char *prop_name, const void *prop_data,
-		int *prop_data_sz)
+static void module_fwupd_push_error_status(void)
 {
 	int err;
-	unsigned int sz = 4 * 1024;
-
-	sz = ((*prop_data_sz + FIT_RSP_MARGIN_SIZE_IN_KB) > FIT_DFL_RSP_SIZE_IN_KB) ?
-		(*prop_data_sz + FIT_RSP_MARGIN_SIZE_IN_KB) : FIT_DFL_RSP_SIZE_IN_KB;
-
-	err = fit_init_resp(res_fit, sz);
-	if (err)
-		goto fit_init_err;
-
-	err = fdt_property_u32(res_fit, HL_FIT_LOAD_TYPE_PROP, load_type);
-	if (err)
-		goto fit_init_err;
-
-	err = fdt_property(res_fit, prop_name, prop_data, *prop_data_sz);
-	if (err)
-		goto fit_init_err;
-
-	err = fit_fin_resp(res_fit, &sz);
-	if (err)
-		goto fit_init_err;
-
-	*prop_data_sz = sz;
-
-fit_init_err:
-	return err;
-}
-
-static void fwupd_get_proto_status(struct hl_upd_proto_status *status, int err, int serr)
-{
-	status->version = HL_UPD_PROTO_VERSION;
-	status->compat_mask = cpu_boot_dev_sts_get_all();
-	status->fw_comp = HL_FW_COMP_PREBOOT;
-	status->err = err;
-	status->sub_err = serr;
-}
-
-static int module_fwupd_proto_status(void)
-{
+	void *offset = state.map_offset;
 	struct hl_upd_proto_status status;
-	int size = sizeof(status);
-	int err;
 
-	state.map_offset = (void *)FIT_IMG_BASE;
-
-	fwupd_get_proto_status(&status, 0, 0);
-
-	err = fwupd_create_resp(state.map_offset, FIT_LOAD_TYPE_PROTO_STATUS,
-				HL_FIT_PROTO_STATUS_PROP, &status, &size);
-	state.map_size = size;
-
+	err = module_fwupd_proto_status(&status);
 	if (err)
-		preboot_puts("Failed to create status FIT image\n");
+		offset = 0;
 
-	return err;
+	module_fwupd_send_status(status.err, offset);
 }
 
 void module_fwupd_run(void)
@@ -418,6 +493,7 @@
 
 	if (state.sending_error_flag && preboot_error_is_pending()) {
 		preboot_puts("FW UPD prepare ERROR_STATUS\n");
+		module_fwupd_push_error_status();
 		state.sending_error_flag = false;
 	}
 
@@ -428,7 +504,7 @@
 		module_fwupd_send_status(sts, NULL);
 		break;
 	case HL_UPD_GET_PROTO_STATUS:
-		err = module_fwupd_proto_status();
+		err = module_fwupd_proto_status(NULL);
 		if (err)
 			module_fwupd_send_status(HL_UPD_STS_ERR, NULL);
 		else
